spec/integration/epoch_interruption_spec.rb

```diff
@@ -0,0 +1,83 @@
+module Wasmtime
+  RSpec.describe "Epoch interruption" do
```

Comment:
What happens when the trap occurs when we are inside Ruby code? I imagine the VM would not be very happy if we were to interrupt it before it had a chance to make sure it is is a good state.

I'm wondering if we may want to use something like `rb_postponed_job_register_one` to ensure signal safety, and give Ruby a chance to finish up whatever it was trying to do?

Reply:
Maybe we want an [InterruptHandle](https://docs.rs/wasmtime/0.19.0/wasmtime/struct.InterruptHandle.html) which does dirty work.

Either way, let’s add some nasty edge case tests for this. Off the top of my head, we need to test:

- calling into a Ruby host method that just `sleep(60)`
- calling into ruby host method that’s generating a ton of garbage in an infinite loop, `with_gc_stress`. Run that about 100 times to make sure we experience an interrupt during GC

Reply:
> Hmm, that link is for wasmtime 0.19, looks like it got removed in https://github.com/bytecodealliance/wasmtime/pull/3925. From the changelog:

Indeed, this was removed in favour of epoch based interruption.

Reply:
FWIW I don't think that having a test that verifies that function imports are not interrupted makes a lot of sense. We are testing the internals of the engine (Wasmtime) at that point, more than anything else. All the interruption mechanisms(fuel and epoch) are concerned about the execution of the Wasm code, they are not concerned about the function imports provided by the embedding. More concretely, epoch checks are performed on [entering each Wasm function](https://github.com/bytecodealliance/wasmtime/pull/3699/files#diff-f4e65f6eab2ed0d2568b98f7090476dab5cf37bbf7e7d5ef10a6a2ca9e97c33eR635), so if the deadline has passed after executing a function import, the interruption should happen at that point in the program.

Reply:
That's a fair point -- we are definitely testing the wasmtime contract here. I'm tempted to leave the test in given it's "cheap" and asserts the behaviour doesn't change on us which would have pretty bad consequences. But if you feel strongly about it I'm happy to remove it too, no strong opinon.

At any rate I'll merge this but it's trivial to remove that commit after, so do not hesitate.